src/libraries/System.Net.Connections/src/System/Net/Connections/ConnectionBase.cs

```diff
@@ -0,0 +1,78 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.Net.Connections
+{
+    /// <summary>
+    /// Provides base functionality shared between singular (e.g. TCP) and multiplexed (e.g. QUIC) connections.
+    /// </summary>
+    public abstract class ConnectionBase : IDisposable, IAsyncDisposable
+    {
+        private bool _disposed;
+
+        /// <summary>
+        /// Properties exposed by this connection.
+        /// </summary>
+        public abstract IConnectionProperties ConnectionProperties { get; }
+
+        /// <summary>
+        /// The local endpoint of this connection, if any.
+        /// </summary>
+        public abstract EndPoint? LocalEndPoint { get; }
+
+        /// <summary>
+        /// The remote endpoint of this connection, if any.
+        /// </summary>
+        public abstract EndPoint? RemoteEndPoint { get; }
+
+        /// <summary>
+        /// Closes the connection.
+        /// </summary>
+        /// <param name="method">The method to use when closing the connection.</param>
+        /// <param name="cancellationToken">A cancellation token for the asynchronous operation.</param>
+        /// <returns>A <see cref="ValueTask"/> for the asynchronous operation.</returns>
+        public async ValueTask CloseAsync(ConnectionCloseMethod method = ConnectionCloseMethod.GracefulShutdown, CancellationToken cancellationToken = default)
+        {
+            if (!_disposed)
+            {
+                await CloseAsyncCore(method, cancellationToken).ConfigureAwait(false);
+                _disposed = true;
+            }
+            GC.SuppressFinalize(this);
+        }
+
+        /// <summary>
+        /// Closes the connection.
+        /// </summary>
+        /// <param name="method">The method to use when closing the connection.</param>
+        /// <param name="cancellationToken">A cancellation token for the asynchronous operation.</param>
+        /// <returns>A <see cref="ValueTask"/> for the asynchronous operation.</returns>
+        protected abstract ValueTask CloseAsyncCore(ConnectionCloseMethod method, CancellationToken cancellationToken);
+
+        /// <summary>
+        /// Disposes of the connection.
+        /// </summary>
+        /// <remarks>This is equivalent to calling <see cref="CloseAsync(ConnectionCloseMethod, CancellationToken)"/> with the method <see cref="ConnectionCloseMethod.Immediate"/>.</remarks>
+        public void Dispose()
+        {
+            ValueTask t = CloseAsync(ConnectionCloseMethod.GracefulShutdown, CancellationToken.None);
+
+            if (t.IsCompleted) t.GetAwaiter().GetResult();
+            else t.AsTask().GetAwaiter().GetResult();
```

Comment:
Do we need the AsTask?

Reply:
`ValueTask` does not support synchronously waiting, so we do need a `Task`.